File: swift/bugzero-kata/bugzero-kata/Game.swift
//
//  Game.swift
//  bugzero-kata
//
//  Created by Frédéric Ruaudel on 20/04/2018.
//

import Foundation

class Game {
    var players = [String]()
    var places = [Int]()
    var purses = [Int]()
    var inPenaltyBox = [Bool]()
    
    var popQuestions = [String]()
    var scienceQuestions = [String]()
    var sportsQuestions = [String]()
    var rockQuestions = [String]()
    
    var currentPlayer: Int = 0
    var isGettingOutOfPenaltyBox: Bool = false
    
    init() {
        for i in 0..<50 {
            popQuestions.append("Pop question \(i)")
            scienceQuestions.append("Science question \(i)")
            sportsQuestions.append("Sports question \(i)")
            rockQuestions.append(createRockQuestions(index: i))
        }
    }
    
    func createRockQuestions(index: Int) -> String {
        return "Rock question \(index)"
    }
    
    func isPlayable() -> Bool {
        return (howManyPlayers() >= 2)
    }
    
    func add(_ playerName: String) -> Bool {
        players.append(playerName)
        places.append(0)
        purses.append(0)
        inPenaltyBox.append(false)
        
        print("\(playerName) was added")
        print("They are player number \(players.count)")
        
        return true
    }
    
    func howManyPlayers() -> Int {
        return players.count
    }
    
    func roll(_ roll: Int) {
        print("\(players[currentPlayer]) is the current player")
        print("They have rolled a \(roll)")
        
        if inPenaltyBox[currentPlayer] {
            if roll % 2 != 0 {
                isGettingOutOfPenaltyBox = true
                print("\(players[currentPlayer]) is getting out of the penalty box")
                movePlayerAndAskQuestion(roll)
            } else {
                print("\(players[currentPlayer]) is not getting out of the penalty box")
                isGettingOutOfPenaltyBox = false
            }
        } else {
            movePlayerAndAskQuestion(roll)
        }
    }
    
    private func movePlayerAndAskQuestion(_ roll: Int) {
        places[currentPlayer] = places[currentPlayer] + roll
        if places[currentPlayer] > 11 { places[currentPlayer] = places[currentPlayer] - 12 }
        
        print("\(players[currentPlayer])'s new location is \(places[currentPlayer])")
        print("The category is \(currentCategory())")
        
        askQuestion()
    }
    
    private func askQuestion() {
        if (currentCategory() == "Pop") {
            print(popQuestions.remove(at: 0))
        }
        if (currentCategory() == "Science") {
            print(scienceQuestions.remove(at: 0))
        }
        if (currentCategory() == "Sports") {
            print(sportsQuestions.remove(at: 0))
        }
        if (currentCategory() == "Rock") {
            print(rockQuestions.remove(at: 0))
        }
    }
    
    private func currentCategory() -> String {
        if places[currentPlayer] == 0 { return "Pop" };
        if places[currentPlayer] == 4 { return "Pop" };
        if places[currentPlayer] == 8 { return "Pop" };
        if places[currentPlayer] == 1 { return "Science" };
        if places[currentPlayer] == 5 { return "Science" };
        if places[currentPlayer] == 9 { return "Science" };
        if places[currentPlayer] == 2 { return "Sports" };
        if places[currentPlayer] == 6 { return "Sports" };
        if places[currentPlayer] == 10 { return "Sports" };
        return "Rock";
    }
    
    func wasCorrectlyAnswered() -> Bool {
        if inPenaltyBox[currentPlayer] {
            if isGettingOutOfPenaltyBox {
                print("Answer was correct!!!!")
                currentPlayer += 1
                if (currentPlayer == players.count) { currentPlayer = 0 }
                purses[currentPlayer] += 1
                print("\(players[currentPlayer]) now has \(purses[currentPlayer]) Gold Coins.")
                let winner = didPlayerWin()
                
                return winner
            } else {
                currentPlayer += 1
                if (currentPlayer == players.count) { currentPlayer = 0 }
                return true
            }
        } else {
            print("Answer was correct!!!!")
            purses[currentPlayer] += 1
            print("\(players[currentPlayer]) now has \(purses[currentPlayer]) Gold Coins.")
            let winner = didPlayerWin()
            currentPlayer += 1
            if (currentPlayer == players.count) { currentPlayer = 0 }
            
            return winner
        }
    }
    
    func wrongAnswer() -> Bool {
        print("Question was incorrectly answered")
        print("\(players[currentPlayer]) was sent to the penalty box")
        inPenaltyBox[currentPlayer] = true
        
        currentPlayer += 1
        if (currentPlayer == players.count) { currentPlayer = 0 }
        return true
    }
    
    private func didPlayerWin() -> Bool {
        return !(purses[currentPlayer] == 6)
    }
}

File: scala/src/main/scala/com/adaptionsoft/games/uglytrivia/Game.scala
package com.adaptionsoft.games.uglytrivia

import scala.collection.mutable

class Game() {
  val players = mutable.ListBuffer[String]()
  val places = new Array[Int](6)
  val purses = new Array[Int](6)
  val inPenaltyBox = new Array[Boolean](6)
  val popQuestions = mutable.ListBuffer[String]()
  val scienceQuestions = mutable.ListBuffer[String]()
  val sportsQuestions = mutable.ListBuffer[String]()
  val rockQuestions = mutable.ListBuffer[String]()
  var currentPlayer = 0
  var isGettingOutOfPenaltyBox: Boolean = false

  (0 until 50).foreach { i =>
    popQuestions.append("Pop Question " + i)
    scienceQuestions.append("Science Question " + i)
    sportsQuestions.append("Sports Question " + i)
    rockQuestions.append(createRockQuestion(i))
  }

  def createRockQuestion(index: Int): String = "Rock Question " + index

  def isPlayable: Boolean = howManyPlayers >= 2

  def add(playerName: String): Boolean = {
    players.append(playerName)
    places(howManyPlayers) = 0
    purses(howManyPlayers) = 0
    inPenaltyBox(howManyPlayers) = false
    println(playerName + " was added")
    println("They are player number " + players.size)
    true
  }

  def howManyPlayers: Int = players.size

  def roll(roll: Int): Unit = {
    println(players(currentPlayer) + " is the current player")
    println("They have rolled a " + roll)
    if (inPenaltyBox(currentPlayer)) if (roll % 2 != 0) {
      isGettingOutOfPenaltyBox = true
      println(players(currentPlayer) + " is getting out of the penalty box")
      movePlayerAndAskQuestion(roll)
    } else {
      println(players(currentPlayer) + " is not getting out of the penalty box")
      isGettingOutOfPenaltyBox = false
    } else
      movePlayerAndAskQuestion(roll)
  }

  private def movePlayerAndAskQuestion(roll: Int): Unit = {
    places(currentPlayer) = places(currentPlayer) + roll
    if (places(currentPlayer) > 11) places(currentPlayer) = places(currentPlayer) - 12
    println(players(currentPlayer) + "'s new location is " + places(currentPlayer))
    println("The category is " + currentCategory)
    askQuestion()
  }

  private def askQuestion(): Unit = {
    if (currentCategory eq "Pop") println(popQuestions.remove(0))
    if (currentCategory eq "Science") println(scienceQuestions.remove(0))
    if (currentCategory eq "Sports") println(sportsQuestions.remove(0))
    if (currentCategory eq "Rock") println(rockQuestions.remove(0))
  }

  private def currentCategory: String = {
    if (places(currentPlayer) == 0) return "Pop"
    if (places(currentPlayer) == 4) return "Pop"
    if (places(currentPlayer) == 8) return "Pop"
    if (places(currentPlayer) == 1) return "Science"
    if (places(currentPlayer) == 5) return "Science"
    if (places(currentPlayer) == 9) return "Science"
    if (places(currentPlayer) == 2) return "Sports"
    if (places(currentPlayer) == 6) return "Sports"
    if (places(currentPlayer) == 10) return "Sports"
    "Rock"
  }

  def wasCorrectlyAnswered: Boolean =
    if (inPenaltyBox(currentPlayer)) {
      if (isGettingOutOfPenaltyBox) {
        println("Answer was correct!!!!")
        currentPlayer += 1
        if (currentPlayer == players.size) currentPlayer = 0
        purses(currentPlayer) += 1
        println(players(currentPlayer) + " now has " + purses(currentPlayer) + " Gold Coins.")
        val winner = didPlayerWin
        winner
      } else {
        currentPlayer += 1
        if (currentPlayer == players.size) currentPlayer = 0
        true
      }
    } else {
      println("Answer was corrent!!!!")
      purses(currentPlayer) += 1
      println(players(currentPlayer) + " now has " + purses(currentPlayer) + " Gold Coins.")
      val winner = didPlayerWin
      currentPlayer += 1
      if (currentPlayer == players.size) currentPlayer = 0
      winner
    }

  def wrongAnswer: Boolean = {
    println("Question was incorrectly answered")
    println(players(currentPlayer) + " was sent to the penalty box")
    inPenaltyBox(currentPlayer) = true
    currentPlayer += 1
    if (currentPlayer == players.size) currentPlayer = 0
    true
  }


  private def didPlayerWin: Boolean = !(purses(currentPlayer) == 6)
}
